File: backend/src/models/Message.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IMessage extends Document {
  content: string;
  senderId: mongoose.Types.ObjectId;
  senderUsername: string;
  roomId: mongoose.Types.ObjectId;
  timestamp: Date;
  edited: boolean;
  editedAt?: Date;
}

const messageSchema = new Schema<IMessage>({
  content: {
    type: String,
    required: [true, 'Message content is required'],
    trim: true,
    maxlength: [1000, 'Message cannot exceed 1000 characters']
  },
  senderId: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  senderUsername: {
    type: String,
    required: true
  },
  roomId: {
    type: Schema.Types.ObjectId,
    ref: 'ChatRoom',
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  },
  edited: {
    type: Boolean,
    default: false
  },
  editedAt: {
    type: Date
  }
});

// Indexes for better query performance
messageSchema.index({ roomId: 1, timestamp: -1 });
messageSchema.index({ senderId: 1 });

const Message = mongoose.model<IMessage>('Message', messageSchema);
export default Message;

File: frontend/lib/models/message.dart
class Message {
  final String id;
  final String content;
  final String senderId;
  final String senderUsername;
  final String roomId;
  final DateTime timestamp;

  Message({
    required this.id,
    required this.content,
    required this.senderId,
    required this.senderUsername,
    required this.roomId,
    required this.timestamp,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    try {
      final message = Message(
        id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
        content: json['content']?.toString() ?? '',
        senderId: json['senderId']?.toString() ?? json['sender_id']?.toString() ?? '',
        senderUsername: json['senderUsername']?.toString() ?? json['sender_username']?.toString() ?? '',
        roomId: json['roomId']?.toString() ?? json['room_id']?.toString() ?? '',
        timestamp: json['timestamp'] != null 
            ? DateTime.parse(json['timestamp'].toString())
            : DateTime.now(),
      );
      return message;
    } catch (e) {
      rethrow;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'content': content,
      'senderId': senderId,
      'senderUsername': senderUsername,
      'roomId': roomId,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}